### PR TITLE
Update the onnx Gemm op to FC/FCTransposed logic in caffe2 onnx backend

### PR DESCRIPTION
**Pull request by JerryShih**
The broadcast is used by default when the opset version is greater then 6.
#9874

### CODE REVIEW

**Comment by JerryShih** [comment]
@houseroad @yinghai
How about this update?

**Comment by JerryShih** [comment]
@houseroad 
There are some shape mismatch bugs in my patch. Please skip this reviewing request.
I will update for these bugs soon.

**Comment by JerryShih** [comment]
@houseroad 
Could you please review these patch?

File: caffe2/python/onnx/tests/c2_ref_test.py
Comment (houseroad) [inline caffe2/python/onnx/tests/c2_ref_test.py]: nit: an empty line

File: caffe2/python/onnx/tests/c2_ref_test.py
Comment (houseroad) [inline caffe2/python/onnx/tests/c2_ref_test.py]: transB=1

File: caffe2/python/onnx/tests/c2_ref_test.py
Comment (houseroad) [inline caffe2/python/onnx/tests/c2_ref_test.py]: transB=1

File: caffe2/python/onnx/tests/c2_ref_test.py
Comment (houseroad) [inline caffe2/python/onnx/tests/c2_ref_test.py]: broadcast=1

**Review by houseroad — APPROVED** [review]
Overall looks pretty good.

Nit: use int instead of bool when calling make_node, since `transA`, `transB`, and `broadcast` are all `int`.

**Comment by JerryShih** [comment]
@houseroad 
Thank you. All boolean values are converted to int.
Wait for the CI test again.

**Comment by JerryShih** [comment]
@houseroad 
CI checks have passed

File: caffe2/onnx/backend.cc
Comment (bddppq) [inline caffe2/onnx/backend.cc]: If this check doesn't pass, it means the onnx model is *invalid*, so it's not just a matter of whether using fc or not, instead here should raise error.

File: caffe2/onnx/backend.cc
Comment (bddppq) [inline caffe2/onnx/backend.cc]: This new code will only try to use FC when input_b's shape information is available, which is not desired.

File: caffe2/python/onnx/tests/c2_ref_test.py
Comment (houseroad) [inline caffe2/python/onnx/tests/c2_ref_test.py]: This case may be incorrect, because C has to match A*B's dimensions. Let's set C's shape as (3,2) too.

**Review by bddppq — DISMISSED** [review]
I think what this PR is trying to improve is to use FCTransposed when trans_b is False, this is a good idea. Could you update the pull request's title & summary accordingly?

File: caffe2/onnx/backend.cc
Comment (houseroad) [reply caffe2/onnx/backend.cc]: It's possible... if c is one-element tensor.

File: caffe2/onnx/backend.cc
Comment (houseroad) [reply caffe2/onnx/backend.cc]: I think in real world models, input b and c are usually weights, so we should be able to optimize them.

Comment (bddppq) [reply caffe2/onnx/backend.cc]: no gemm is pretty common, so you don't know whether a, b and c are inputs or weights.

File: caffe2/onnx/backend.cc
Comment (bddppq) [reply caffe2/onnx/backend.cc]: @houseroad in that case, explicitly check that

File: caffe2/onnx/backend.cc
Comment (houseroad) [reply caffe2/onnx/backend.cc]: @JerryShih let's be more aggressive, two things to do:
1) report error if the invalid condition appear,
2) if B's shape information is not available, always assume shapes match.

File: caffe2/onnx/backend.cc
Comment (houseroad) [reply caffe2/onnx/backend.cc]: Very legit. @JerryShih could you address the above comments?

**Comment by houseroad** [comment]
A potential follow up change (after getting this merged):


Right now, FC only accpet 1d bias, however, we can improve FC's implementation to accept 2d bias.

FC's implementation:
https://github.com/pytorch/pytorch/blob/master/caffe2/operators/fully_connected_op.h#L69

Same to FCTranspose.

When FC accepts 2d bias, we can always turn a Gemm into Scale + FC(Transpose)

File: caffe2/python/onnx/tests/c2_ref_test.py
Comment (JerryShih) [reply caffe2/python/onnx/tests/c2_ref_test.py]: checked.

File: caffe2/onnx/backend.cc
Comment (JerryShih) [reply caffe2/onnx/backend.cc]: How about these two case if c is a scalar?
a)

        A = np.random.randn(1, 3).astype(np.float32)
        B = np.random.randn(3, 1).astype(np.float32)
        C = np.random.randn(1).astype(np.float32)
        node_def = make_node(
            'Gemm',
            ['A', 'B', 'C'],
            ["Y"])
        output = c2.run_node(node_def, [A, B, C])
        np.testing.assert_almost_equal(output["Y"], np.dot(A, B) + C)

b)

        A = np.random.randn(3, 2).astype(np.float32)
        B = np.random.randn(2, 4).astype(np.float32)
        C = np.random.randn(1).astype(np.float32)
        node_def = make_node(
            'Gemm',
            ['A', 'B', 'C'],
            ["Y"])
        output = c2.run_node(node_def, [A, B, C])
        np.testing.assert_almost_equal(output["Y"], np.dot(A, B) + C)

The a) and b) could both use "MatMul+add".
The a) can use FC, but b) can't.
If there is no B's shape information, we can't recognize which case we have now.

@bddppq @houseroad

File: caffe2/onnx/backend.cc
Comment (JerryShih) [reply caffe2/onnx/backend.cc]: I still can't catch the idea. If there is no shape info, we can't make sure the input c fit the FC op's requirement(just like the scalar c case in previous comment). In this case, I think we should just use "MatMul+add". 
What did I miss?

@bddppq @houseroad

**Comment by JerryShih** [comment]
@houseroad @bddppq 
I update some comments in review page.

**Comment by JerryShih** [comment]
@houseroad @bddppq
What do you think for my comments? Did I miss something?

File: caffe2/onnx/backend.cc
Comment (houseroad) [reply caffe2/onnx/backend.cc]: If only C's shape information is available and C is a 1d tensor, but B's shape is not available, we can aggressively assume that C's shape matches B's last dimension (or first dimension if transB). 1-element tensor is special, you can handle it with `MatMul + Add`. The point here is that `FC`'s performance is much better than `MatMul + Add`, so we would like to use `FC` more aggressively.

File: caffe2/onnx/backend.cc
Comment (houseroad) [reply caffe2/onnx/backend.cc]: If B's shape information is not available, treating scalar as `MatMul+Add` is fine.

**Comment by JerryShih** [comment]
@houseroad @bddppq 
Request review again.

I update the condition for converting to FC.

1. If c is "1d" tensor and c is not a scalar, use FC.
2. If c is a scalar and b's shape is not available, skip FC.
3. If c is a scalar and a*b's last dim is 1, use FC.
4. If c is a scalar and a*b's last dim is not 1, skip FC.

In most case, we will use FC.

**Comment by JerryShih** [comment]
@houseroad @bddppq 
Any feedback?

"pr/caffe2-py2-cuda9.0-cudnn7-windows-build — Build failed"
This test always failed, but that's not related to these patches.

**Comment by houseroad** [comment]
I have restarted the build, don’t worry. I will review today later.

**Comment by JerryShih** [comment]
@houseroad
It already has taken 3 days waiting for merging pr. Is it any test failed?

**Comment by houseroad** [comment]
I am running some internal test cases. Will update it soon.

**Comment by JerryShih** [comment]
yes, I will do that.

**Comment by houseroad** [comment]
Awesome, if you have any question, feel free to ping on gitter.

**Pull request closed without merging**
